app/models/site_statistic.rb

```diff
@@ -203,67 +337,206 @@ def self.users_stats(at_time = Time.now)
+      size: 0,
+      track_total_hits: true,
+      filters: [
+        range: {
+          created_at: {
+            gte: at_time - 1.days,
+            lte: at_time
+          }
+        }
+      ]
+    ).total_entries
+    { count: count,
+      last_7_days: last_7_days,
+      today: today
     }
   end
 
   def self.taxa_stats(at_time = Time.now)
     at_time = at_time.utc
-    { species_counts: Taxon.of_rank_equiv_or_lower(10).joins(:observations).
-        where("observations.created_at <= ?", at_time).
-        distinct.
-        count(:id),
-      species_counts_by_site: Hash[ Taxon.joins(observations: :site).
-        where("observations.created_at <= ?", at_time).
-        select("sites.name, count(distinct taxa.id) as count").
-        of_rank_equiv_or_lower(10).
-        group("sites.id").
-        order( Arel.sql( "count(distinct taxa.id) desc" ) ).
-        collect{ |a|
-          [ a["name"], a["count"].to_i ]
+    species_counts = Observation.elastic_search(
+      size: 0,
+      track_total_hits: true,
+      filters: [
+        {
+          range: {
+            created_at: {
+              lte: at_time
+            }
+          }
+        },
+        {
+          range: {
+            "taxon.rank_level": {
+              lte: 10
+            }
+          }
         }
       ],
-      count_by_rank: Hash[ Taxon.joins(:observations).
-        where("observations.created_at <= ?", at_time).
-        where("taxa.rank_level > 0").
-        select("taxa.rank_level, count(distinct observations.id) as count").
-        group("taxa.rank_level").
-        order( Arel.sql( "count(distinct observations.id) desc" ) ).
-        collect{ |a| [
-          Taxon::RANK_LEVELS.select{ |k,v| v == a["rank_level"] }.first.try(:first) || "other",
-          a["count"].to_i
-        ]}
-      ]
+      aggregate: {
+        species_counts: {
+          cardinality: {
+            field: "taxon.id"
+          }
+        }
+      }
+    ).response.aggregations.species_counts.value    
+    sites = Site.select(:id, :name).map{ |s| 
+        [ s.id, s.name ]
+      }.to_h
+    species_counts_by_site = Observation.elastic_search(
+      size: 0,
+      track_total_hits: true,
+      filters: [
+        {
+          range: {
+            created_at: {
+              lte: at_time
+            }
+          }
+        },
+        {
+          range: {
+            "taxon.rank_level": {
+              lte: 10
+            }
+          }
+        }
+      ],
+      aggregate: {
+        group_by_sites: {
+          terms: {
+            field: "site_id",
+            size: 1000
+          },
+          aggs: {
+            distinct_taxon_id: {
+              cardinality: {
+                field: "taxon.id"
+              }
+            }
+          }
+        }
+      }
+    ).response.aggregations.group_by_sites.
+      buckets.map do |b| 
+          if sites.key?(b["key"])
+            site_name = sites[b["key"]]
+            [ site_name, b["doc_count"] ]
```

Comment:
This is using `doc_count` and not the value of the `distinct_taxon_id` aggregation, so it reflects observation counts and not distinct taxon counts

Reply:
Fixed